core/src/main/java/io/questdb/cutlass/auth/AuthUtils.java

```diff
@@ -24,19 +24,115 @@
 
 package io.questdb.cutlass.auth;
 
+import io.questdb.std.CharSequenceObjHashMap;
 import io.questdb.std.Chars;
+import io.questdb.std.ThreadLocal;
 
+import java.io.BufferedInputStream;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.InputStreamReader;
 import java.math.BigInteger;
+import java.nio.ByteBuffer;
 import java.security.*;
 import java.security.interfaces.ECKey;
 import java.security.spec.*;
 import java.util.Base64;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 public final class AuthUtils {
     public static final String EC_ALGORITHM = "EC";
     public static final String EC_CURVE = "secp256r1";
+    // there are 2 types of signatures:
+    // 1. sigP1363 format is 64 bytes
+    // 2. sigDER format is 72 bytes
+    public static final int MAX_SIGNATURE_LENGTH = 72;
+    // (72 bytes / 3) * 4 = 96
+    public static final int MAX_SIGNATURE_LENGTH_BASE64 = 96;
     public static final String SIGNATURE_TYPE_DER = "SHA256withECDSA";
     public static final String SIGNATURE_TYPE_P1363 = "SHA256withECDSAinP1363Format";
+    private static final Pattern TOKEN_PATTERN = Pattern.compile("\\s*(\\S+)(.*)");
+
+    private static final ThreadLocal<Signature> tlSigDER = new ThreadLocal<>(() -> {
+        try {
+            return Signature.getInstance(AuthUtils.SIGNATURE_TYPE_DER);
+        } catch (NoSuchAlgorithmException ex) {
+            throw new Error(ex);
+        }
+    });
+    private static final ThreadLocal<Signature> tlSigP1363 = new ThreadLocal<>(() -> {
+        try {
+            return Signature.getInstance(AuthUtils.SIGNATURE_TYPE_P1363);
+        } catch (NoSuchAlgorithmException ex) {
+            throw new Error(ex);
+        }
+    });
+
+    public static boolean isSignatureMatch(PublicKey publicKey, byte[] challenge, ByteBuffer signature) throws InvalidKeyException, SignatureException {
+        Signature sig = signature.remaining() == 64 ? tlSigP1363.get() : tlSigDER.get();
+        // On some out of date JDKs zeros can be valid signature because of a bug in the JDK code
+        // Check that it's not the case.
```

Comment:
Is it fine that checkAllZeros() traverses buffer from 0 to remaining() while 
sig.verify uses bytes from position() to remaining() ?

Reply:
good one, thanks!